### 1. Reading dataset and Importing the libraries

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import os
os.chdir('E:/Data Science/Competitions/Projects/Poverty Level Prediction')

In [2]:
df = pd.read_csv('costa_rica.csv')
df.head()

Id      v2a1  hacdor  rooms  hacapo  v14a  refrig  v18q  v18q1  \
0  ID_279628684  190000.0       0      3       0     1       1     0    NaN   
1  ID_f29eb3ddd  135000.0       0      4       0     1       1     1    1.0   
2  ID_68de51c94       NaN       0      8       0     1       1     0    NaN   
3  ID_d671db89c  180000.0       0      5       0     1       1     1    1.0   
4  ID_d56d6f5f5  180000.0       0      5       0     1       1     1    1.0   

   r4h1  ...  SQBescolari  SQBage  SQBhogar_total  SQBedjefe  SQBhogar_nin  \
0     0  ...          100    1849               1        100             0   
1     0  ...          144    4489               1        144             0   
2     0  ...          121    8464               1          0             0   
3     0  ...           81     289              16        121             4   
4     0  ...          121    1369              16        121             4   

   SQBovercrowding  SQBdependency  SQBmeaned  agesq  Target  
0         1.000000            0.0      100.0   1849       4  
1         1.000000           64.0      144.0   4489       4  
2         0.250000           64.0      121.0   8464       4  
3         1.777778            1.0      121.0    289       4  
4         1.777778            1.0      121.0   1369       4  

[5 rows x 143 columns]

### 2. Understanding the data

In [3]:
df.shape

(9557, 143)

In [4]:
## dropped id column as it won't be useful for prediction
ndf = df.drop(['Id'],axis = 1)
ndf.head(2)

v2a1  hacdor  rooms  hacapo  v14a  refrig  v18q  v18q1  r4h1  r4h2  \
0  190000.0       0      3       0     1       1     0    NaN     0     1   
1  135000.0       0      4       0     1       1     1    1.0     0     1   

   ...  SQBescolari  SQBage  SQBhogar_total  SQBedjefe  SQBhogar_nin  \
0  ...          100    1849               1        100             0   
1  ...          144    4489               1        144             0   

   SQBovercrowding  SQBdependency  SQBmeaned  agesq  Target  
0              1.0            0.0      100.0   1849       4  
1              1.0           64.0      144.0   4489       4  

[2 rows x 142 columns]

In [6]:
## Renaming the columns for ease of understanding

ndf.columns = ['monthly_rent','overcrowd_bed','rooms','overcrowd_room',
 'has_bathroom','has_fridge','has_tablet','n_tablets','male_under_12',
'male_above_12','n_males','female_under_12',
'female_above_12','n_females','n_under_12','n_above_12','n_persons',
'house_size','total_persons','schooling_years','schooling_years_behind',
'size_house','wall_material_type1','wall_material_type2','wall_material_type3','wall_material_type4',
'wall_material_type5','wall_material_type6','wall_material_type7','wall_material_type8',
'floor_material_type1','floor_material_type2','floor_material_type3','floor_material_type4',
'has_floor','floor_material_type5', 'roof_material_type1',
'roof_material_type2','roof_material_type3',
'roof_material_type4','has_ceiling','water_provision_type1',
'water_provision_type2','has_water_provision','electricity_provider_type1',
'electricity_provider_type2','has_electricity','electricity_provider_type3',
'has_toilet','toilet_type2','toilet_type3','toilet_type4','toilet_type5',
'has_kitchen','energy_source_type2','energy_source_type3','energy_source_type4',
'truck_garbage_disposal','garbage_disposal_type2','garbage_disposal_type3',
'garbage_disposal_type4','garbage_disposal_type5','garbage_disposal_type6',
'wall_quality1','wall_quality2','is_wall_quality_good','roof_quality1','roof_quality2',
'is_roof_quality_good','floor_quality1','floor_quality2','is_floor_quality_good',
'disabled_person','is_male','is_female','is_under_10','is_free_coupled',
'is_married','is_divorced','is_separated','is_widow','is_single',
'is_house_head','is_partner','is_son_daughter','is_stepson_son_daughter',
'is_s/d_inlaw','is_grandchild','is_parent','is_f/m_inlaw','is_bro_sis',
'is_bro/sis_inlaw','is_other','is_non_family','house_id','n_children',
'n_adults','n_above_65','total_individuals','dependency_rate','male_years_of_education',
'female_years_of_education','adults_years_of_education',
 'has_edu_level','edu_prim_inc', 'edu_prim_comp','edu_asec_inc','edu_asec_comp',
'edu_tsec_inc','edu_tsec_comp','edu_ug_above','edu_pg','n_bed',
'n_persons_per_room','is_owned','house_own_type2','is_rented',
'house_own_type4','house_own_type5','has_computer','has_tv','has_mobile','n_mobiles',
'region_central','region_chorotega','region_pac_central','region_Brunca',
'region_huetar_atl','region_huetar_nor','is_urban','is_rural','age',
'SQBescolari','SQBage','SQBhogar_total','SQBedjefe',
'SQBhogar_nin','SQBovercrowding','SQBdependency','SQBmeaned', 
'agesq','Target'
]

In [7]:
ndf.head()

monthly_rent  overcrowd_bed  rooms  overcrowd_room  has_bathroom  \
0      190000.0              0      3               0             1   
1      135000.0              0      4               0             1   
2           NaN              0      8               0             1   
3      180000.0              0      5               0             1   
4      180000.0              0      5               0             1   

   has_fridge  has_tablet  n_tablets  male_under_12  male_above_12  ...  \
0           1           0        NaN              0              1  ...   
1           1           1        1.0              0              1  ...   
2           1           0        NaN              0              0  ...   
3           1           1        1.0              0              2  ...   
4           1           1        1.0              0              2  ...   

   SQBescolari  SQBage  SQBhogar_total  SQBedjefe  SQBhogar_nin  \
0          100    1849               1        100             0   
1          144    4489               1        144             0   
2          121    8464               1          0             0   
3           81     289              16        121             4   
4          121    1369              16        121             4   

   SQBovercrowding  SQBdependency  SQBmeaned  agesq  Target  
0         1.000000            0.0      100.0   1849       4  
1         1.000000           64.0      144.0   4489       4  
2         0.250000           64.0      121.0   8464       4  
3         1.777778            1.0      121.0    289       4  
4         1.777778            1.0      121.0   1369       4  

[5 rows x 142 columns]

In [8]:
ndf.tail()

monthly_rent  overcrowd_bed  rooms  overcrowd_room  has_bathroom  \
9552       80000.0              0      6               0             1   
9553       80000.0              0      6               0             1   
9554       80000.0              0      6               0             1   
9555       80000.0              0      6               0             1   
9556       80000.0              0      6               0             1   

      has_fridge  has_tablet  n_tablets  male_under_12  male_above_12  ...  \
9552           1           0        NaN              0              2  ...   
9553           1           0        NaN              0              2  ...   
9554           1           0        NaN              0              2  ...   
9555           1           0        NaN              0              2  ...   
9556           1           0        NaN              0              2  ...   

      SQBescolari  SQBage  SQBhogar_total  SQBedjefe  SQBhogar_nin  \
9552           81    2116              25         81             1   
9553            0       4              25         81             1   
9554           25    2500              25         81             1   
9555          121     676              25         81             1   
9556           64     441              25         81             1   

      SQBovercrowding  SQBdependency  SQBmeaned  agesq  Target  
9552           1.5625         0.0625    68.0625   2116       2  
9553           1.5625         0.0625    68.0625      4       2  
9554           1.5625         0.0625    68.0625   2500       2  
9555           1.5625         0.0625    68.0625    676       2  
9556           1.5625         0.0625    68.0625    441       2  

[5 rows x 142 columns]

In [9]:
ndf.describe()

monthly_rent  overcrowd_bed        rooms  overcrowd_room  has_bathroom  \
count  2.697000e+03    9557.000000  9557.000000     9557.000000   9557.000000   
mean   1.652316e+05       0.038087     4.955530        0.023648      0.994768   
std    1.504571e+05       0.191417     1.468381        0.151957      0.072145   
min    0.000000e+00       0.000000     1.000000        0.000000      0.000000   
25%    8.000000e+04       0.000000     4.000000        0.000000      1.000000   
50%    1.300000e+05       0.000000     5.000000        0.000000      1.000000   
75%    2.000000e+05       0.000000     6.000000        0.000000      1.000000   
max    2.353477e+06       1.000000    11.000000        1.000000      1.000000   

        has_fridge   has_tablet    n_tablets  male_under_12  male_above_12  \
count  9557.000000  9557.000000  2215.000000    9557.000000    9557.000000   
mean      0.957623     0.231767     1.404063       0.385895       1.559171   
std       0.201459     0.421983     0.763131       0.680779       1.036574   
min       0.000000     0.000000     1.000000       0.000000       0.000000   
25%       1.000000     0.000000     1.000000       0.000000       1.000000   
50%       1.000000     0.000000     1.000000       0.000000       1.000000   
75%       1.000000     0.000000     2.000000       1.000000       2.000000   
max       1.000000     1.000000     6.000000       5.000000       8.000000   

       ...  SQBescolari       SQBage  SQBhogar_total    SQBedjefe  \
count  ...  9557.000000  9557.000000     9557.000000  9557.000000   
mean   ...    74.222769  1643.774302       19.132887    53.500262   
std    ...    76.777549  1741.197050       18.751395    78.445804   
min    ...     0.000000     0.000000        1.000000     0.000000   
25%    ...    16.000000   289.000000        9.000000     0.000000   
50%    ...    36.000000   961.000000       16.000000    36.000000   
75%    ...   121.000000  2601.000000       25.000000    81.000000   
max    ...   441.000000  9409.000000      169.000000   441.000000   

       SQBhogar_nin  SQBovercrowding  SQBdependency    SQBmeaned        agesq  \
count   9557.000000      9557.000000    9557.000000  9552.000000  9557.000000   
mean       3.844826         3.249485       3.900409   102.588867  1643.774302   
std        6.946296         4.129547      12.511831    93.516890  1741.197050   
min        0.000000         0.040000       0.000000     0.000000     0.000000   
25%        0.000000         1.000000       0.111111    36.000000   289.000000   
50%        1.000000         2.250000       0.444444    81.000000   961.000000   
75%        4.000000         4.000000       1.777778   134.560010  2601.000000   
max       81.000000        36.000000      64.000000  1369.000000  9409.000000   

            Target  
count  9557.000000  
mean      3.302292  
std       1.009565  
min       1.000000  
25%       3.000000  
50%       4.000000  
75%       4.000000  
max       4.000000  

[8 rows x 138 columns]

## Creating a new dataframe for ease of understanding

In [ ]:
ndf = ndf.drop(['SQBescolari','SQBage','SQBhogar_total','SQBedjefe','SQBhogar_nin','SQBovercrowding',
          'SQBdependency','SQBmeaned','agesq'], axis = 1)

In [15]:
print(ndf.columns.tolist())

['monthly_rent', 'overcrowd_bed', 'rooms', 'overcrowd_room', 'has_bathroom', 'has_fridge', 'has_tablet', 'n_tablets', 'male_under_12', 'male_above_12', 'n_males', 'female_under_12', 'female_above_12', 'n_females', 'n_under_12', 'n_above_12', 'n_persons', 'house_size', 'total_persons', 'schooling_years', 'schooling_years_behind', 'size_house', 'wall_material_type1', 'wall_material_type2', 'wall_material_type3', 'wall_material_type4', 'wall_material_type5', 'wall_material_type6', 'wall_material_type7', 'wall_material_type8', 'floor_material_type1', 'floor_material_type2', 'floor_material_type3', 'floor_material_type4', 'has_floor', 'floor_material_type5', 'roof_material_type1', 'roof_material_type2', 'roof_material_type3', 'roof_material_type4', 'has_ceiling', 'water_provision_type1', 'water_provision_type2', 'has_water_provision', 'electricity_provider_type1', 'electricity_provider_type2', 'has_electricity', 'electricity_provider_type3', 'has_toilet', 'toilet_type2', 'toilet_type3', 'to

In [17]:
### Working on house own type

ndf[['is_owned','house_own_type2','house_own_type4','house_own_type5','is_rented']]

is_owned  house_own_type2  house_own_type4  house_own_type5  is_rented
0            0                0                0                0          1
1            0                0                0                0          1
2            1                0                0                0          0
3            0                0                0                0          1
4            0                0                0                0          1
...        ...              ...              ...              ...        ...
9552         0                0                0                0          1
9553         0                0                0                0          1
9554         0                0                0                0          1
9555         0                0                0                0          1
9556         0                0                0                0          1

[9557 rows x 5 columns]

In [18]:
### creating a new feature as 'ownership'

own_house = ['is_owned','is_rented','house_own_type2','house_own_type4','house_own_type5']

ndf['ownership'] = ndf[own_house].idxmax(axis = 1)

ndf['ownership'] = ndf['ownership'].map({'is_owned':'own','is_rented':'rented',
                                         'house_own_type2':'own','house_own_type4':'other','house_own_type5':'other'})

In [128]:
ndf['ownership'].value_counts()

own       6872
rented    1736
other      949
Name: ownership, dtype: int64

In [24]:
### creating new feature 'region'

regions = ['region_central','region_chorotega','region_pac_central','region_Brunca',
'region_huetar_atl','region_huetar_nor']

ndf['region'] = ndf[regions].idxmax(axis = 1)

ndf['region'] = ndf['region'].map({'region_central':'central','region_chorotega':'chorotega','region_pac_central':'pac_central',
                                   'region_Brunca':'Brunca','region_huetar_atl':'huetar_alt','region_huetar_nor':'huetar_nor'})

In [25]:
### creating a new feature 'zone' to represent urban or rural area

ndf['zone'] = ndf['is_urban'].apply(lambda x: 'urban' if x == 1 else 'rural')

In [26]:
### creating new feature 'gender' to represent male and female

ndf['gender'] = ndf['is_male'].apply(lambda x: 'male' if x == 1 else 'female')

In [27]:
## Walls
walls = ['wall_material_type1','wall_material_type2','wall_material_type3','wall_material_type4',
         'wall_material_type5','wall_material_type6','wall_material_type7','wall_material_type8']

ndf['wall_material_type'] = ndf[walls].idxmax(axis = 1)


In [31]:
### creating a new feature 'wall_material_type' so that the number of features get reduced

ndf['wall_material_type'] = ndf['wall_material_type'].apply(lambda x: 'brick' if x == 'wall_material_type1' 
                                                            else ('cement' if x == 'wall_material_type3' else 'other'))

In [32]:
ndf['wall_material_type'].value_counts()

brick     5677
other     2083
cement    1797
Name: wall_material_type, dtype: int64

In [49]:
## Floor
floors = ['floor_material_type1','floor_material_type2','floor_material_type3','floor_material_type4',
          'floor_material_type5']

ndf['floor_material_type'] = ndf[floors].idxmax(axis = 1)


In [50]:
ndf.loc[ ndf['has_floor'] == 1, 'floor_material_type'] = 'no'

In [52]:
ndf['floor_material_type']=ndf['floor_material_type'].apply(lambda x: 'ceramic' if x == 'floor_material_type1' 
                                                            else ('cement' if x == 'floor_material_type2'
                                                            else ( x if x == 'no' else 'other')))

In [53]:
ndf['floor_material_type'].value_counts()

ceramic    6621
cement     2130
other       649
no          157
Name: floor_material_type, dtype: int64

In [62]:
## Roof
roofs = ['roof_material_type1','roof_material_type2','roof_material_type3','roof_material_type4']

ndf['roof_material_type'] = ndf[roofs].idxmax(axis = 1)


In [65]:
ndf['roof_material_type'] = ndf['roof_material_type'].apply(lambda x: 'metal' if x == 'roof_material_type1' 
                                                            else ( 'cement' if x == 'roof_material_type2' 
                                                            else'other'))

In [66]:
ndf['roof_material_type'].value_counts()

metal     9337
cement     169
other       51
Name: roof_material_type, dtype: int64

In [75]:
## Water
water_provision = ['water_provision_type1','water_provision_type2']

ndf['water_provision_type'] = ndf[water_provision].idxmax(axis = 1)


In [78]:
ndf.loc[ ndf['has_water_provision'] == 1, 'water_provision_type'] = 'no'

In [80]:
ndf['water_provision_type']=ndf['water_provision_type'].apply(lambda x: 'inside' if x == 'water_provision_type1' 
                                                              else ('outside' if x == 'water_provision_type2' 
                                                              else x))

In [81]:
ndf['water_provision_type'].value_counts()

inside     9220
outside     303
no           34
Name: water_provision_type, dtype: int64

In [96]:
## Electricity

el_provider = ['electricity_provider_type1',
'electricity_provider_type2','electricity_provider_type3']

ndf['elec_provider_type'] = ndf[el_provider].idxmax(axis = 1)



In [97]:
ndf.loc[ ndf['has_electricity'] == 1, 'elec_provider_type'] = 'no'


In [100]:

ndf['elec_provider_type']=ndf['elec_provider_type'].apply(lambda x: 'cooperative' if x == 'electricity_provider_type3' 
                                                              else ('other' if x == 'electricity_provider_type1' or x == 'electricity_provider_type2'
                                                              else x))

In [101]:
ndf['elec_provider_type'].value_counts()

other          8477
cooperative    1059
no               21
Name: elec_provider_type, dtype: int64

In [106]:

## Toilet
toilets = ['toilet_type2','toilet_type3','toilet_type4','toilet_type5']

ndf['toilet_type'] = ndf[toilets].idxmax(axis = 1)


In [108]:
ndf.loc[ ndf['has_toilet'] == 1, 'toilet_type'] = 'no'


In [110]:
ndf['toilet_type'] = ndf['toilet_type'].apply(lambda x: 'septic' if x == 'toilet_type3'
                                              else (x if x == 'no'
                                              else 'other'))


In [111]:
ndf['toilet_type'].value_counts()

septic    7304
other     2216
no          37
Name: toilet_type, dtype: int64

In [112]:

## garbage disposal
disposal = ['garbage_disposal_type2','garbage_disposal_type3','garbage_disposal_type4',
            'garbage_disposal_type5','garbage_disposal_type6']

ndf['garbage_disposal_type'] = ndf[disposal].idxmax(axis = 1)



In [114]:
ndf['garbage_disposal_type'] = ndf['garbage_disposal_type'].apply(lambda x: 'buried' if x == 'garbage_disposal_type2' 
                                                                  else 'other')

In [115]:
ndf['garbage_disposal_type'].value_counts()

buried    8723
other      834
Name: garbage_disposal_type, dtype: int64

In [122]:

## Energy Source
ene_source = ['energy_source_type2','energy_source_type3','energy_source_type4']

ndf['energy_source_type'] = ndf[ene_source].idxmax(axis = 1)


In [123]:
ndf.loc[ ndf['has_kitchen'] == 1, 'energy_source_type'] = 'no'

In [125]:
ndf['energy_source_type'] = ndf['energy_source_type'].apply(lambda x: 'elec' if x == 'energy_source_type2' 
                                                            else (x if x == 'no'
                                                            else 'other'))

In [126]:
ndf['energy_source_type'].value_counts()

other    4860
elec     4679
no         18
Name: energy_source_type, dtype: int64

In [129]:
### Dropping the unnecessary columns

to_drop = ['wall_material_type1','wall_material_type2','wall_material_type3','wall_material_type4',
'wall_material_type5','wall_material_type6','wall_material_type7','wall_material_type8',
'floor_material_type1','floor_material_type2','floor_material_type3','floor_material_type4',
'floor_material_type5', 'roof_material_type1',
'roof_material_type2','roof_material_type3',
'roof_material_type4','water_provision_type1',
'water_provision_type2','electricity_provider_type1',
'electricity_provider_type2','electricity_provider_type3',
'toilet_type2','toilet_type3','toilet_type4','toilet_type5',
'energy_source_type2','energy_source_type3','energy_source_type4',
'garbage_disposal_type2','garbage_disposal_type3',
'garbage_disposal_type4','garbage_disposal_type5','garbage_disposal_type6',
'wall_quality1','wall_quality2','roof_quality1','roof_quality2',
'floor_quality1','floor_quality2',
'house_own_type2','house_own_type4','house_own_type5',
'region_central','region_chorotega','region_pac_central','region_Brunca',
'region_huetar_atl','region_huetar_nor','is_urban','is_rural','is_owned','is_rented'
]

In [130]:
ndf = ndf.drop(to_drop, axis = 1)

In [131]:
ndf.shape

(9557, 92)

In [133]:
print(ndf.columns.tolist())

['monthly_rent', 'overcrowd_bed', 'rooms', 'overcrowd_room', 'has_bathroom', 'has_fridge', 'has_tablet', 'n_tablets', 'male_under_12', 'male_above_12', 'n_males', 'female_under_12', 'female_above_12', 'n_females', 'n_under_12', 'n_above_12', 'n_persons', 'house_size', 'total_persons', 'schooling_years', 'schooling_years_behind', 'size_house', 'has_floor', 'has_ceiling', 'has_water_provision', 'has_electricity', 'has_toilet', 'has_kitchen', 'truck_garbage_disposal', 'is_wall_quality_good', 'is_roof_quality_good', 'is_floor_quality_good', 'disabled_person', 'is_male', 'is_female', 'is_under_10', 'is_free_coupled', 'is_married', 'is_divorced', 'is_separated', 'is_widow', 'is_single', 'is_house_head', 'is_partner', 'is_son_daughter', 'is_stepson_son_daughter', 'is_s/d_inlaw', 'is_grandchild', 'is_parent', 'is_f/m_inlaw', 'is_bro_sis', 'is_bro/sis_inlaw', 'is_other', 'is_non_family', 'house_id', 'n_children', 'n_adults', 'n_above_65', 'total_individuals', 'dependency_rate', 'male_years_of_e

In [134]:
## Education 

education = ['edu_prim_inc', 'edu_prim_comp', 'edu_asec_inc', 'edu_asec_comp', 
             'edu_tsec_inc', 'edu_tsec_comp']

In [135]:
ndf['edu_primary'] = ndf['edu_prim_comp']

In [136]:
ndf['edu_sec'] = ndf['edu_asec_comp']
ndf['edu_techsec'] = ndf['edu_tsec_comp']

In [137]:
ndf = ndf.drop(education, axis = 1)
ndf.head()

monthly_rent  overcrowd_bed  rooms  overcrowd_room  has_bathroom  \
0      190000.0              0      3               0             1   
1      135000.0              0      4               0             1   
2           NaN              0      8               0             1   
3      180000.0              0      5               0             1   
4      180000.0              0      5               0             1   

   has_fridge  has_tablet  n_tablets  male_under_12  male_above_12  ...  \
0           1           0        NaN              0              1  ...   
1           1           1        1.0              0              1  ...   
2           1           0        NaN              0              0  ...   
3           1           1        1.0              0              2  ...   
4           1           1        1.0              0              2  ...   

   floor_material_type  roof_material_type  water_provision_type  \
0              ceramic              cement                inside   
1                other               metal                inside   
2              ceramic               metal                inside   
3              ceramic               metal                inside   
4              ceramic               metal                inside   

   elec_provider_type  toilet_type  garbage_disposal_type  energy_source_type  \
0               other        other                 buried               other   
1               other        other                 buried                elec   
2               other        other                 buried                elec   
3               other        other                 buried                elec   
4               other        other                 buried                elec   

   edu_primary  edu_sec  edu_techsec  
0            0        0            0  
1            0        0            0  
2            0        1            0  
3            0        0            0  
4            0        1            0  

[5 rows x 89 columns]

In [138]:
### Dropping these as a feature named 'n_persons' represents the total number of members in the family

ndf = ndf.drop(['total_persons','total_individuals'], axis = 1)
ndf.head()

monthly_rent  overcrowd_bed  rooms  overcrowd_room  has_bathroom  \
0      190000.0              0      3               0             1   
1      135000.0              0      4               0             1   
2           NaN              0      8               0             1   
3      180000.0              0      5               0             1   
4      180000.0              0      5               0             1   

   has_fridge  has_tablet  n_tablets  male_under_12  male_above_12  ...  \
0           1           0        NaN              0              1  ...   
1           1           1        1.0              0              1  ...   
2           1           0        NaN              0              0  ...   
3           1           1        1.0              0              2  ...   
4           1           1        1.0              0              2  ...   

   floor_material_type  roof_material_type  water_provision_type  \
0              ceramic              cement                inside   
1                other               metal                inside   
2              ceramic               metal                inside   
3              ceramic               metal                inside   
4              ceramic               metal                inside   

   elec_provider_type  toilet_type  garbage_disposal_type  energy_source_type  \
0               other        other                 buried               other   
1               other        other                 buried                elec   
2               other        other                 buried                elec   
3               other        other                 buried                elec   
4               other        other                 buried                elec   

   edu_primary  edu_sec  edu_techsec  
0            0        0            0  
1            0        0            0  
2            0        1            0  
3            0        0            0  
4            0        1            0  

[5 rows x 87 columns]

In [139]:
print(ndf.columns.tolist())

['monthly_rent', 'overcrowd_bed', 'rooms', 'overcrowd_room', 'has_bathroom', 'has_fridge', 'has_tablet', 'n_tablets', 'male_under_12', 'male_above_12', 'n_males', 'female_under_12', 'female_above_12', 'n_females', 'n_under_12', 'n_above_12', 'n_persons', 'house_size', 'schooling_years', 'schooling_years_behind', 'size_house', 'has_floor', 'has_ceiling', 'has_water_provision', 'has_electricity', 'has_toilet', 'has_kitchen', 'truck_garbage_disposal', 'is_wall_quality_good', 'is_roof_quality_good', 'is_floor_quality_good', 'disabled_person', 'is_male', 'is_female', 'is_under_10', 'is_free_coupled', 'is_married', 'is_divorced', 'is_separated', 'is_widow', 'is_single', 'is_house_head', 'is_partner', 'is_son_daughter', 'is_stepson_son_daughter', 'is_s/d_inlaw', 'is_grandchild', 'is_parent', 'is_f/m_inlaw', 'is_bro_sis', 'is_bro/sis_inlaw', 'is_other', 'is_non_family', 'house_id', 'n_children', 'n_adults', 'n_above_65', 'dependency_rate', 'male_years_of_education', 'female_years_of_education'

In [140]:
### Combining the features 'overcrowd_bed' and 'overcrowd_room' to 
### a single feature as 'is_overcrowded' which is binary with values 0 and 1

ndf['is_overcrowded'] = ndf['overcrowd_bed'] + ndf['overcrowd_room']

In [141]:
ndf['is_overcrowded'] = ndf['is_overcrowded'].apply(lambda x: 1 if x > 1 else x)

In [142]:
ndf['is_overcrowded'].value_counts()

0    9157
1     400
Name: is_overcrowded, dtype: int64

In [143]:
ndf = ndf.drop(['overcrowd_bed','overcrowd_room'], axis = 1)
ndf.head()

monthly_rent  rooms  has_bathroom  has_fridge  has_tablet  n_tablets  \
0      190000.0      3             1           1           0        NaN   
1      135000.0      4             1           1           1        1.0   
2           NaN      8             1           1           0        NaN   
3      180000.0      5             1           1           1        1.0   
4      180000.0      5             1           1           1        1.0   

   male_under_12  male_above_12  n_males  female_under_12  ...  \
0              0              1        1                0  ...   
1              0              1        1                0  ...   
2              0              0        0                0  ...   
3              0              2        2                1  ...   
4              0              2        2                1  ...   

   roof_material_type  water_provision_type  elec_provider_type  toilet_type  \
0              cement                inside               other        other   
1               metal                inside               other        other   
2               metal                inside               other        other   
3               metal                inside               other        other   
4               metal                inside               other        other   

   garbage_disposal_type  energy_source_type  edu_primary  edu_sec  \
0                 buried               other            0        0   
1                 buried                elec            0        0   
2                 buried                elec            0        1   
3                 buried                elec            0        0   
4                 buried                elec            0        1   

   edu_techsec  is_overcrowded  
0            0               0  
1            0               0  
2            0               0  
3            0               0  
4            0               0  

[5 rows x 86 columns]

In [144]:
print(ndf.columns.tolist())

['monthly_rent', 'rooms', 'has_bathroom', 'has_fridge', 'has_tablet', 'n_tablets', 'male_under_12', 'male_above_12', 'n_males', 'female_under_12', 'female_above_12', 'n_females', 'n_under_12', 'n_above_12', 'n_persons', 'house_size', 'schooling_years', 'schooling_years_behind', 'size_house', 'has_floor', 'has_ceiling', 'has_water_provision', 'has_electricity', 'has_toilet', 'has_kitchen', 'truck_garbage_disposal', 'is_wall_quality_good', 'is_roof_quality_good', 'is_floor_quality_good', 'disabled_person', 'is_male', 'is_female', 'is_under_10', 'is_free_coupled', 'is_married', 'is_divorced', 'is_separated', 'is_widow', 'is_single', 'is_house_head', 'is_partner', 'is_son_daughter', 'is_stepson_son_daughter', 'is_s/d_inlaw', 'is_grandchild', 'is_parent', 'is_f/m_inlaw', 'is_bro_sis', 'is_bro/sis_inlaw', 'is_other', 'is_non_family', 'house_id', 'n_children', 'n_adults', 'n_above_65', 'dependency_rate', 'male_years_of_education', 'female_years_of_education', 'adults_years_of_education', 'has

In [145]:
ndf.shape

(9557, 86)

### 3. Check for missing values and treating them

In [146]:
ndf.isnull().sum()[ndf.isnull().sum()>0]

monthly_rent                 6860
n_tablets                    7342
schooling_years_behind       7928
adults_years_of_education       5
dtype: int64

In [147]:
## n_Tablets

ndf[ndf['n_tablets'].isnull()].head()


monthly_rent  rooms  has_bathroom  has_fridge  has_tablet  n_tablets  \
0      190000.0      3             1           1           0        NaN   
2           NaN      8             1           1           0        NaN   
7      130000.0      2             1           1           0        NaN   
8      130000.0      2             1           1           0        NaN   
9      130000.0      2             1           1           0        NaN   

   male_under_12  male_above_12  n_males  female_under_12  ...  \
0              0              1        1                0  ...   
2              0              0        0                0  ...   
7              0              1        1                2  ...   
8              0              1        1                2  ...   
9              0              1        1                2  ...   

   roof_material_type  water_provision_type  elec_provider_type  toilet_type  \
0              cement                inside               other        other   
2               metal                inside               other        other   
7               metal                inside               other        other   
8               metal                inside               other        other   
9               metal                inside               other        other   

   garbage_disposal_type  energy_source_type  edu_primary  edu_sec  \
0                 buried               other            0        0   
2                 buried                elec            0        1   
7                 buried               other            0        0   
8                 buried               other            0        0   
9                 buried               other            0        1   

   edu_techsec  is_overcrowded  
0            0               0  
2            0               0  
7            0               1  
8            0               1  
9            0               1  

[5 rows x 86 columns]

In [148]:
ndf[ndf['n_tablets'].isnull()]['has_tablet'].unique()

array([0], dtype=int64)

In [149]:
ndf['n_tablets'] = ndf['has_tablet'].fillna(0)

In [150]:
ndf[ndf['n_tablets'].isnull()]

Empty DataFrame
Columns: [monthly_rent, rooms, has_bathroom, has_fridge, has_tablet, n_tablets, male_under_12, male_above_12, n_males, female_under_12, female_above_12, n_females, n_under_12, n_above_12, n_persons, house_size, schooling_years, schooling_years_behind, size_house, has_floor, has_ceiling, has_water_provision, has_electricity, has_toilet, has_kitchen, truck_garbage_disposal, is_wall_quality_good, is_roof_quality_good, is_floor_quality_good, disabled_person, is_male, is_female, is_under_10, is_free_coupled, is_married, is_divorced, is_separated, is_widow, is_single, is_house_head, is_partner, is_son_daughter, is_stepson_son_daughter, is_s/d_inlaw, is_grandchild, is_parent, is_f/m_inlaw, is_bro_sis, is_bro/sis_inlaw, is_other, is_non_family, house_id, n_children, n_adults, n_above_65, dependency_rate, male_years_of_education, female_years_of_education, adults_years_of_education, has_edu_level, edu_ug_above, edu_pg, n_bed, n_persons_per_room, has_computer, has_tv, has_mobile, n_mobiles, age, Target, ownership, region, zone, gender, wall_material_type, floor_material_type, roof_material_type, water_provision_type, elec_provider_type, toilet_type, garbage_disposal_type, energy_source_type, edu_primary, edu_sec, edu_techsec, is_overcrowded]
Index: []

[0 rows x 86 columns]

In [151]:
## Adults years of education

ndf[ndf['adults_years_of_education'].isnull()]

monthly_rent  rooms  has_bathroom  has_fridge  has_tablet  n_tablets  \
1291           NaN      7             1           1           0          0   
1840      110000.0      1             1           1           0          0   
1841      110000.0      1             1           1           0          0   
2049      180000.0      3             1           1           0          0   
2050      180000.0      3             1           1           0          0   

      male_under_12  male_above_12  n_males  female_under_12  ...  \
1291              0              0        0                0  ...   
1840              0              1        1                0  ...   
1841              0              1        1                0  ...   
2049              0              2        2                0  ...   
2050              0              2        2                0  ...   

      roof_material_type  water_provision_type  elec_provider_type  \
1291               metal                inside               other   
1840               metal                inside               other   
1841               metal                inside               other   
2049               metal                inside               other   
2050               metal                inside               other   

      toilet_type  garbage_disposal_type  energy_source_type  edu_primary  \
1291        other                 buried                elec            0   
1840       septic                 buried               other            1   
1841       septic                 buried               other            0   
2049        other                 buried                elec            0   
2050        other                 buried                elec            0   

      edu_sec  edu_techsec  is_overcrowded  
1291        0            0               0  
1840        0            0               0  
1841        0            0               0  
2049        0            1               0  
2050        0            1               0  

[5 rows x 86 columns]

In [152]:
ndf[['house_id','age','n_adults','schooling_years', 'adults_years_of_education']].head()

house_id  age  n_adults  schooling_years  adults_years_of_education
0  21eb7fcc1   43         1               10                       10.0
1  0e5d7a658   67         1               12                       12.0
2  2c7317ea8   92         1               11                       11.0
3  2b58d945f   17         2                9                       11.0
4  2b58d945f   37         2               11                       11.0

In [153]:
ndf[ndf['adults_years_of_education'].isnull()][['house_id','age','n_adults','schooling_years','adults_years_of_education']]

house_id  age  n_adults  schooling_years  adults_years_of_education
1291  1b31fd159   18         0               10                        NaN
1840  a874b7ce7   18         0                6                        NaN
1841  a874b7ce7   18         0                4                        NaN
2049  faaebf71a   19         0               12                        NaN
2050  faaebf71a   19         0               12                        NaN

In [154]:
for i in ndf[ndf['adults_years_of_education'].isnull()]['house_id'].unique():
    ndf.loc[(ndf['house_id'] == i) & (ndf['age'] >= 18), 'adults_years_of_education'] = ndf[ndf['house_id'] == i]['schooling_years'].sum()/ndf[ndf['house_id'] == i].shape[0]

In [155]:
ndf[ndf['adults_years_of_education'].isnull()]

Empty DataFrame
Columns: [monthly_rent, rooms, has_bathroom, has_fridge, has_tablet, n_tablets, male_under_12, male_above_12, n_males, female_under_12, female_above_12, n_females, n_under_12, n_above_12, n_persons, house_size, schooling_years, schooling_years_behind, size_house, has_floor, has_ceiling, has_water_provision, has_electricity, has_toilet, has_kitchen, truck_garbage_disposal, is_wall_quality_good, is_roof_quality_good, is_floor_quality_good, disabled_person, is_male, is_female, is_under_10, is_free_coupled, is_married, is_divorced, is_separated, is_widow, is_single, is_house_head, is_partner, is_son_daughter, is_stepson_son_daughter, is_s/d_inlaw, is_grandchild, is_parent, is_f/m_inlaw, is_bro_sis, is_bro/sis_inlaw, is_other, is_non_family, house_id, n_children, n_adults, n_above_65, dependency_rate, male_years_of_education, female_years_of_education, adults_years_of_education, has_edu_level, edu_ug_above, edu_pg, n_bed, n_persons_per_room, has_computer, has_tv, has_mobile, n_mobiles, age, Target, ownership, region, zone, gender, wall_material_type, floor_material_type, roof_material_type, water_provision_type, elec_provider_type, toilet_type, garbage_disposal_type, energy_source_type, edu_primary, edu_sec, edu_techsec, is_overcrowded]
Index: []

[0 rows x 86 columns]

In [156]:
## schooling years behind

ndf[ndf['schooling_years_behind'].isnull()].head()

monthly_rent  rooms  has_bathroom  has_fridge  has_tablet  n_tablets  \
0      190000.0      3             1           1           0          0   
1      135000.0      4             1           1           1          1   
2           NaN      8             1           1           0          0   
4      180000.0      5             1           1           1          1   
5      180000.0      5             1           1           1          1   

   male_under_12  male_above_12  n_males  female_under_12  ...  \
0              0              1        1                0  ...   
1              0              1        1                0  ...   
2              0              0        0                0  ...   
4              0              2        2                1  ...   
5              0              2        2                1  ...   

   roof_material_type  water_provision_type  elec_provider_type  toilet_type  \
0              cement                inside               other        other   
1               metal                inside               other        other   
2               metal                inside               other        other   
4               metal                inside               other        other   
5               metal                inside               other        other   

   garbage_disposal_type  energy_source_type  edu_primary  edu_sec  \
0                 buried               other            0        0   
1                 buried                elec            0        0   
2                 buried                elec            0        1   
4                 buried                elec            0        1   
5                 buried                elec            0        1   

   edu_techsec  is_overcrowded  
0            0               0  
1            0               0  
2            0               0  
4            0               0  
5            0               0  

[5 rows x 86 columns]

In [157]:
ndf[ndf['schooling_years_behind'].isnull()][['has_edu_level','schooling_years','edu_primary','edu_sec','edu_techsec']]

has_edu_level  schooling_years  edu_primary  edu_sec  edu_techsec
0                 0               10            0        0            0
1                 0               12            0        0            0
2                 0               11            0        1            0
4                 0               11            0        1            0
5                 0               11            0        1            0
...             ...              ...          ...      ...          ...
9552              0                9            0        0            0
9553              1                0            0        0            0
9554              0                5            0        0            0
9555              0               11            0        1            0
9556              0                8            0        0            0

[7928 rows x 5 columns]

In [158]:
ndf[(ndf['schooling_years_behind'].isnull()) & (ndf['has_edu_level']==1)][['has_edu_level','schooling_years','edu_primary','edu_sec','edu_techsec','edu_ug_above','edu_pg']]

has_edu_level  schooling_years  edu_primary  edu_sec  edu_techsec  \
35                1                0            0        0            0   
67                1                0            0        0            0   
86                1                0            0        0            0   
96                1                0            0        0            0   
116               1                0            0        0            0   
...             ...              ...          ...      ...          ...   
9539              1                0            0        0            0   
9540              1                0            0        0            0   
9549              1                0            0        0            0   
9551              1                0            0        0            0   
9553              1                0            0        0            0   

      edu_ug_above  edu_pg  
35               0       0  
67               0       0  
86               0       0  
96               0       0  
116              0       0  
...            ...     ...  
9539             0       0  
9540             0       0  
9549             0       0  
9551             0       0  
9553             0       0  

[1183 rows x 7 columns]

In [159]:
ndf.loc[(ndf['schooling_years_behind'].isnull()) & (ndf['has_edu_level']==1), 'schooling_years_behind'] = 0

In [160]:
ndf[(ndf['schooling_years_behind'].isnull()) & (ndf['has_edu_level']==1)][['has_edu_level','schooling_years','edu_primary','edu_sec','edu_techsec','edu_ug_above','edu_pg']]

Empty DataFrame
Columns: [has_edu_level, schooling_years, edu_primary, edu_sec, edu_techsec, edu_ug_above, edu_pg]
Index: []

In [161]:
ndf[ndf['schooling_years_behind'].isnull()].shape

(6745, 86)

In [162]:
ndf[(ndf['schooling_years_behind'].isnull()) & (ndf['has_edu_level']==0)][['age','has_edu_level','schooling_years',
                                'schooling_years_behind', 'is_free_coupled', 'is_married', 'is_divorced', 'is_separated', 
                                'is_widow', 'is_single']]

age  has_edu_level  schooling_years  schooling_years_behind  \
0      43              0               10                     NaN   
1      67              0               12                     NaN   
2      92              0               11                     NaN   
4      37              0               11                     NaN   
5      38              0               11                     NaN   
...   ...            ...              ...                     ...   
9550   61              0                6                     NaN   
9552   46              0                9                     NaN   
9554   50              0                5                     NaN   
9555   26              0               11                     NaN   
9556   21              0                8                     NaN   

      is_free_coupled  is_married  is_divorced  is_separated  is_widow  \
0                   0           0            1             0         0   
1                   0           0            1             0         0   
2                   0           0            0             0         1   
4                   1           0            0             0         0   
5                   1           0            0             0         0   
...               ...         ...          ...           ...       ...   
9550                0           1            0             0         0   
9552                1           0            0             0         0   
9554                1           0            0             0         0   
9555                0           0            0             0         0   
9556                0           0            0             0         0   

      is_single  
0             0  
1             0  
2             0  
4             0  
5             0  
...         ...  
9550          0  
9552          0  
9554          0  
9555          1  
9556          1  

[6745 rows x 10 columns]

In [164]:
ndf.columns.tolist().index('schooling_years_behind')

17

In [165]:
status = ndf[(ndf['schooling_years_behind'].isnull()) & (ndf['has_edu_level']==0) ][['is_free_coupled', 'is_married', 'is_divorced', 'is_separated', 'is_widow', 'is_single']].idxmax(axis = 1)

In [166]:
ids = ndf[ndf['schooling_years_behind'].isnull()].index

In [167]:
for i,j in zip(ids,status):
    ndf.iloc[i, 17] = ndf[ndf[j] == 1]['schooling_years_behind'].mean()

In [168]:
ndf[ndf['schooling_years_behind'].isnull()]

Empty DataFrame
Columns: [monthly_rent, rooms, has_bathroom, has_fridge, has_tablet, n_tablets, male_under_12, male_above_12, n_males, female_under_12, female_above_12, n_females, n_under_12, n_above_12, n_persons, house_size, schooling_years, schooling_years_behind, size_house, has_floor, has_ceiling, has_water_provision, has_electricity, has_toilet, has_kitchen, truck_garbage_disposal, is_wall_quality_good, is_roof_quality_good, is_floor_quality_good, disabled_person, is_male, is_female, is_under_10, is_free_coupled, is_married, is_divorced, is_separated, is_widow, is_single, is_house_head, is_partner, is_son_daughter, is_stepson_son_daughter, is_s/d_inlaw, is_grandchild, is_parent, is_f/m_inlaw, is_bro_sis, is_bro/sis_inlaw, is_other, is_non_family, house_id, n_children, n_adults, n_above_65, dependency_rate, male_years_of_education, female_years_of_education, adults_years_of_education, has_edu_level, edu_ug_above, edu_pg, n_bed, n_persons_per_room, has_computer, has_tv, has_mobile, n_mobiles, age, Target, ownership, region, zone, gender, wall_material_type, floor_material_type, roof_material_type, water_provision_type, elec_provider_type, toilet_type, garbage_disposal_type, energy_source_type, edu_primary, edu_sec, edu_techsec, is_overcrowded]
Index: []

[0 rows x 86 columns]

In [169]:
## monthly rent

ndf[ndf['monthly_rent'].isnull()].head()

monthly_rent  rooms  has_bathroom  has_fridge  has_tablet  n_tablets  \
2            NaN      8             1           1           0          0   
13           NaN      4             1           1           1          1   
14           NaN      4             1           1           1          1   
26           NaN      5             1           1           0          0   
32           NaN      5             1           1           0          0   

    male_under_12  male_above_12  n_males  female_under_12  ...  \
2               0              0        0                0  ...   
13              0              1        1                0  ...   
14              0              1        1                0  ...   
26              0              1        1                0  ...   
32              0              4        4                0  ...   

    roof_material_type  water_provision_type  elec_provider_type  toilet_type  \
2                metal                inside               other        other   
13               metal                inside               other        other   
14               metal                inside               other        other   
26               metal                inside               other        other   
32               metal                inside               other        other   

    garbage_disposal_type  energy_source_type  edu_primary  edu_sec  \
2                  buried                elec            0        1   
13                 buried                elec            0        0   
14                 buried                elec            0        0   
26                 buried                elec            0        0   
32                 buried                elec            0        0   

    edu_techsec  is_overcrowded  
2             0               0  
13            0               0  
14            0               0  
26            0               0  
32            0               0  

[5 rows x 86 columns]

In [170]:
ndf[ndf['monthly_rent'].isnull()]['region'].value_counts()

central        3758
huetar_alt      750
chorotega       683
Brunca          641
huetar_nor      587
pac_central     441
Name: region, dtype: int64

In [171]:
ndf[ndf['monthly_rent'].isnull()]['zone'].value_counts()

urban    4583
rural    2277
Name: zone, dtype: int64

In [172]:
ndf[ndf['monthly_rent'].isnull()]['ownership'].value_counts()

own      5911
other     949
Name: ownership, dtype: int64

In [173]:
ndf['ownership'].value_counts()

own       6872
rented    1736
other      949
Name: ownership, dtype: int64

In [174]:
ndf.groupby(['region','zone','ownership'])['monthly_rent'].mean()

## fully paid houses are the ones that are owned and do not have any monthly rent so missing values can be replaced by 0

region       zone   ownership
Brunca       rural  other                  NaN
                    own          147615.384615
                    rented        67705.586207
             urban  other                  NaN
                    own          117319.047619
                    rented        71013.513514
central      rural  other                  NaN
                    own          169907.400000
                    rented       120100.333333
             urban  other                  NaN
                    own          236553.988543
                    rented       167623.470588
chorotega    rural  other                  NaN
                    own          324375.000000
                    rented        69059.230769
             urban  other                  NaN
                    own          211983.333333
                    rented       116470.158416
huetar_alt   rural  other                  NaN
                    own          131333.333333
                    rented    

In [176]:
ids = ndf[ndf['monthly_rent'].isnull()].index

reg = ndf[ndf['monthly_rent'].isnull()]['region']

zones = ndf[ndf['monthly_rent'].isnull()]['zone']

for i,r,z in zip(ids,reg,zones):
    ndf.iloc[i, 0] = ndf[((ndf['region'] == r) & (ndf['zone'] == z))]['monthly_rent'].mean()

In [177]:
ndf[ndf['monthly_rent'].isnull()]

Empty DataFrame
Columns: [monthly_rent, rooms, has_bathroom, has_fridge, has_tablet, n_tablets, male_under_12, male_above_12, n_males, female_under_12, female_above_12, n_females, n_under_12, n_above_12, n_persons, house_size, schooling_years, schooling_years_behind, size_house, has_floor, has_ceiling, has_water_provision, has_electricity, has_toilet, has_kitchen, truck_garbage_disposal, is_wall_quality_good, is_roof_quality_good, is_floor_quality_good, disabled_person, is_male, is_female, is_under_10, is_free_coupled, is_married, is_divorced, is_separated, is_widow, is_single, is_house_head, is_partner, is_son_daughter, is_stepson_son_daughter, is_s/d_inlaw, is_grandchild, is_parent, is_f/m_inlaw, is_bro_sis, is_bro/sis_inlaw, is_other, is_non_family, house_id, n_children, n_adults, n_above_65, dependency_rate, male_years_of_education, female_years_of_education, adults_years_of_education, has_edu_level, edu_ug_above, edu_pg, n_bed, n_persons_per_room, has_computer, has_tv, has_mobile, n_mobiles, age, Target, ownership, region, zone, gender, wall_material_type, floor_material_type, roof_material_type, water_provision_type, elec_provider_type, toilet_type, garbage_disposal_type, energy_source_type, edu_primary, edu_sec, edu_techsec, is_overcrowded]
Index: []

[0 rows x 86 columns]

In [178]:
ndf.isnull().sum()[ndf.isnull().sum()>0]

Series([], dtype: int64)

#####  Now, we don't have any missing values

In [179]:
### dropping 'size_house' as it is similar to 'house_size' feature

ndf = ndf.drop('size_house', axis = 1)

##### In the following 3 cells, the features having inconsistent data were treated to make it consistent by replacing yes with 1 and no with 0

In [180]:
### male_years_of_education

ndf['male_years_of_education'] = ndf['male_years_of_education'].apply(lambda x: 1 if x == 'yes' 
                                                                      else (0 if x == 'no' else x))

In [181]:
### female_years_of_education

ndf['female_years_of_education'] = ndf['female_years_of_education'].apply(lambda x: 1 if x == 'yes' 
                                                                          else (0 if x == 'no' else x))

In [183]:
### dependency_rate

ndf['dependency_rate'] = ndf['dependency_rate'].apply(lambda x: 1 if x == 'yes' else (0 if x == 'no' else x))

In [184]:
### dropping house id as it won't contribute in prediction

ndf = ndf.drop(['house_id'], axis = 1)

In [185]:
print(ndf.columns.tolist())

['monthly_rent', 'rooms', 'has_bathroom', 'has_fridge', 'has_tablet', 'n_tablets', 'male_under_12', 'male_above_12', 'n_males', 'female_under_12', 'female_above_12', 'n_females', 'n_under_12', 'n_above_12', 'n_persons', 'house_size', 'schooling_years', 'schooling_years_behind', 'has_floor', 'has_ceiling', 'has_water_provision', 'has_electricity', 'has_toilet', 'has_kitchen', 'truck_garbage_disposal', 'is_wall_quality_good', 'is_roof_quality_good', 'is_floor_quality_good', 'disabled_person', 'is_male', 'is_female', 'is_under_10', 'is_free_coupled', 'is_married', 'is_divorced', 'is_separated', 'is_widow', 'is_single', 'is_house_head', 'is_partner', 'is_son_daughter', 'is_stepson_son_daughter', 'is_s/d_inlaw', 'is_grandchild', 'is_parent', 'is_f/m_inlaw', 'is_bro_sis', 'is_bro/sis_inlaw', 'is_other', 'is_non_family', 'n_children', 'n_adults', 'n_above_65', 'dependency_rate', 'male_years_of_education', 'female_years_of_education', 'adults_years_of_education', 'has_edu_level', 'edu_ug_above

In [186]:
ndf.shape

(9557, 84)

In [187]:
#### Dropping these as represented by other features

ndf = ndf.drop(['is_free_coupled','is_divorced', 'is_separated', 'is_widow', 'is_single',
'is_son_daughter', 'is_stepson_son_daughter', 'is_s/d_inlaw', 'is_grandchild', 
'is_f/m_inlaw', 'is_bro_sis', 'is_bro/sis_inlaw', 'is_other', 'is_non_family'
               ], axis = 1)

In [188]:
print(ndf.columns.tolist())

['monthly_rent', 'rooms', 'has_bathroom', 'has_fridge', 'has_tablet', 'n_tablets', 'male_under_12', 'male_above_12', 'n_males', 'female_under_12', 'female_above_12', 'n_females', 'n_under_12', 'n_above_12', 'n_persons', 'house_size', 'schooling_years', 'schooling_years_behind', 'has_floor', 'has_ceiling', 'has_water_provision', 'has_electricity', 'has_toilet', 'has_kitchen', 'truck_garbage_disposal', 'is_wall_quality_good', 'is_roof_quality_good', 'is_floor_quality_good', 'disabled_person', 'is_male', 'is_female', 'is_under_10', 'is_married', 'is_house_head', 'is_partner', 'is_parent', 'n_children', 'n_adults', 'n_above_65', 'dependency_rate', 'male_years_of_education', 'female_years_of_education', 'adults_years_of_education', 'has_edu_level', 'edu_ug_above', 'edu_pg', 'n_bed', 'n_persons_per_room', 'has_computer', 'has_tv', 'has_mobile', 'n_mobiles', 'age', 'Target', 'ownership', 'region', 'zone', 'gender', 'wall_material_type', 'floor_material_type', 'roof_material_type', 'water_prov

In [189]:
ndf.shape

(9557, 70)

In [190]:
#### dropping these as number of males and females  under and above 12 are represented by 'n_uner_12' and 'n_above_12'

ndf  = ndf.drop(['male_under_12', 'male_above_12','female_under_12', 'female_above_12'], axis = 1)

In [191]:
print(ndf.columns.tolist())

['monthly_rent', 'rooms', 'has_bathroom', 'has_fridge', 'has_tablet', 'n_tablets', 'n_males', 'n_females', 'n_under_12', 'n_above_12', 'n_persons', 'house_size', 'schooling_years', 'schooling_years_behind', 'has_floor', 'has_ceiling', 'has_water_provision', 'has_electricity', 'has_toilet', 'has_kitchen', 'truck_garbage_disposal', 'is_wall_quality_good', 'is_roof_quality_good', 'is_floor_quality_good', 'disabled_person', 'is_male', 'is_female', 'is_under_10', 'is_married', 'is_house_head', 'is_partner', 'is_parent', 'n_children', 'n_adults', 'n_above_65', 'dependency_rate', 'male_years_of_education', 'female_years_of_education', 'adults_years_of_education', 'has_edu_level', 'edu_ug_above', 'edu_pg', 'n_bed', 'n_persons_per_room', 'has_computer', 'has_tv', 'has_mobile', 'n_mobiles', 'age', 'Target', 'ownership', 'region', 'zone', 'gender', 'wall_material_type', 'floor_material_type', 'roof_material_type', 'water_provision_type', 'elec_provider_type', 'toilet_type', 'garbage_disposal_type

In [192]:
ndf.shape

(9557, 66)

In [194]:
ndf = ndf.drop('truck_garbage_disposal', axis = 1)

In [195]:
ndf = ndf.drop('is_under_10', axis = 1)

In [199]:
ndf = ndf.drop('house_size', axis = 1)

In [202]:
ndf = ndf.drop(['is_male','is_female'], axis = 1)

In [203]:
ndf.shape

(9557, 61)

In [204]:
print(ndf.columns.tolist())

['monthly_rent', 'rooms', 'has_bathroom', 'has_fridge', 'has_tablet', 'n_tablets', 'n_males', 'n_females', 'n_under_12', 'n_above_12', 'n_persons', 'schooling_years', 'schooling_years_behind', 'has_floor', 'has_ceiling', 'has_water_provision', 'has_electricity', 'has_toilet', 'has_kitchen', 'is_wall_quality_good', 'is_roof_quality_good', 'is_floor_quality_good', 'disabled_person', 'is_married', 'is_house_head', 'is_partner', 'is_parent', 'n_children', 'n_adults', 'n_above_65', 'dependency_rate', 'male_years_of_education', 'female_years_of_education', 'adults_years_of_education', 'has_edu_level', 'edu_ug_above', 'edu_pg', 'n_bed', 'n_persons_per_room', 'has_computer', 'has_tv', 'has_mobile', 'n_mobiles', 'age', 'Target', 'ownership', 'region', 'zone', 'gender', 'wall_material_type', 'floor_material_type', 'roof_material_type', 'water_provision_type', 'elec_provider_type', 'toilet_type', 'garbage_disposal_type', 'energy_source_type', 'edu_primary', 'edu_sec', 'edu_techsec', 'is_overcrowd

In [205]:
ndf.head()

monthly_rent  rooms  has_bathroom  has_fridge  has_tablet  n_tablets  \
0  190000.000000      3             1           1           0          0   
1  135000.000000      4             1           1           1          1   
2  192412.491466      8             1           1           0          0   
3  180000.000000      5             1           1           1          1   
4  180000.000000      5             1           1           1          1   

   n_males  n_females  n_under_12  n_above_12  ...  roof_material_type  \
0        1          0           0           1  ...              cement   
1        1          0           0           1  ...               metal   
2        0          1           0           1  ...               metal   
3        2          2           1           3  ...               metal   
4        2          2           1           3  ...               metal   

   water_provision_type  elec_provider_type  toilet_type  \
0                inside               other        other   
1                inside               other        other   
2                inside               other        other   
3                inside               other        other   
4                inside               other        other   

   garbage_disposal_type  energy_source_type  edu_primary  edu_sec  \
0                 buried               other            0        0   
1                 buried                elec            0        0   
2                 buried                elec            0        1   
3                 buried                elec            0        0   
4                 buried                elec            0        1   

   edu_techsec  is_overcrowded  
0            0               0  
1            0               0  
2            0               0  
3            0               0  
4            0               0  

[5 rows x 61 columns]

##### Now, we have reduced the number of features from 143 to 61. Also, the data is clean and consistent and ready for further analysis.

In [207]:
### Saving the data for future use

ndf.to_csv('E:\Data Science\Competitions\Projects\Poverty Level Prediction\data.csv', index = False)